File: analysis/revision/scripts/threshold_fig.py
import pandas as pd, numpy as np
from scipy import stats
from sklearn.mixture import GaussianMixture
import matplotlib; matplotlib.use('Agg'); import matplotlib.pyplot as plt

np.random.seed(0)
df = pd.read_csv('./data/Tree_ID_info.csv')
df['site']=np.where(df['plot']=='BGS','BGS','EMS')
sppmap={'rm':'A.rubrum','bg':'N.sylvatica','ro':'Q.rubra','hem':'T.canadensis'}
df['sp']=df['species'].map(sppmap)
ert=pd.read_csv('./data/ERT_application_results.csv')
m=df.merge(ert,on='tree',how='inner').reset_index(drop=True)
metrics=['mean','median','sd','cv','gini','entropy','cma','radialgradient']
Z=m[metrics].copy()
for met in metrics:
    Z[met]=m.groupby('sp')[met].transform(lambda x:(x-x.mean())/x.std(ddof=1))
Zc=Z.values-Z.values.mean(0)
U,S,Vt=np.linalg.svd(Zc,full_matrices=False)
pc=(U*S)[:,0]
if np.corrcoef(pc,m['mean'])[0,1]>0: pc=-pc
m['pc']=pc; m['struct']=m['percent_damaged']>1
mu,med,sd=pc.mean(),np.median(pc),pc.std(ddof=1)
def otsu(x,nb=256):
    h,e=np.histogram(x,bins=nb); c=(e[:-1]+e[1:])/2; w=h/h.sum()
    cw=np.cumsum(w); cm=np.cumsum(w*c); gm=cm[-1]; bt,bv=None,-1
    for i in range(1,nb):
        w0=cw[i]; w1=1-w0
        if w0<=0 or w1<=0: continue
        v=w0*w1*((cm[i]/w0)-((gm-cm[i])/w1))**2
        if v>bv: bv,bt=v,c[i]
    return bt
g2=GaussianMixture(2,random_state=0,n_init=5).fit(pc.reshape(-1,1))
mns=g2.means_.ravel(); o=np.argsort(mns); xs=np.linspace(mns[o[0]],mns[o[1]],2000)
lp=g2._estimate_weighted_log_prob(xs.reshape(-1,1)); d=lp[:,o[0]]-lp[:,o[1]]
idx=np.where(np.diff(np.sign(d))!=0)[0]; t_gmm=xs[idx[0]] if len(idx) else np.nan
t_otsu=otsu(pc)
rules={'mean (published)':(mu,'#c0392b'),'median':(med,'#2980b9'),
       'Otsu break':(t_otsu,'#8e44ad'),'GMM crossover':(t_gmm,'#27ae60')}

fig,(ax1,ax2)=plt.subplots(1,2,figsize=(12.5,4.8))
# Panel A: density
xs2=np.linspace(pc.min()-0.4,pc.max()+0.4,300); kde=stats.gaussian_kde(pc)
ax1.hist(pc,bins=15,density=True,color='0.86',edgecolor='white',zorder=1)
ax1.plot(xs2,kde(xs2),color='k',lw=1.8,zorder=2)
for name,(t,c) in rules.items():
    if np.isfinite(t): ax1.axvline(t,color=c,ls='--',lw=1.7,label=f"{name} = {t:.2f}",zorder=3)
ax1.set_ylim(0,0.42); ax1.set_xlabel('ERT PC1  (higher = lower resistivity / more heterogeneous)')
ax1.set_ylabel('density'); ax1.legend(fontsize=8,frameon=False,loc='upper right')
ax1.set_title('A.  Anomaly axis is unimodal — no natural break\n(BIC favors 1 component over 2)',fontsize=10,loc='left')

# Panel B: sweep threshold, plot BGS vs EMS incipient%
ts=np.linspace(np.percentile(pc,5),np.percentile(pc,95),200)
bgs=[]; ems=[]
mask_bgs=(m.site=='BGS').values; mask_ems=(m.site=='EMS').values; ns=(~m['struct']).values
for t in ts:
    inc=ns&(pc>t)
    bgs.append(inc[mask_bgs].sum()/mask_bgs.sum()*100)
    ems.append(inc[mask_ems].sum()/mask_ems.sum()*100)
ax2.plot(ts,bgs,color='#1f77b4',lw=2.4,label='BGS (wetland)')
ax2.plot(ts,ems,color='#d9822b',lw=2.4,label='EMS (upland)')
ax2.fill_between(ts,ems,bgs,where=np.array(bgs)>=np.array(ems),color='#1f77b4',alpha=0.10)
for name,(t,c) in rules.items():
    if np.isfinite(t): ax2.axvline(t,color=c,ls=':',lw=1.3)
ax2.set_xlabel('anomaly threshold on ERT PC1'); ax2.set_ylabel('% of site trees classified "incipient"')
ax2.legend(fontsize=9,frameon=False,loc='upper right')
ax2.set_title('B.  Wetland > upland incipient holds at EVERY threshold\n(absolute % varies; the direction does not)',fontsize=10,loc='left')
plt.tight_layout(); plt.savefig('.-threshold-fig.png',dpi=175)
print("saved. BGS never below EMS across sweep:", all(np.array(bgs)>=np.array(ems)))
print(f"BGS incipient range across sweep: {min(bgs):.0f}-{max(bgs):.0f}% ; EMS: {min(ems):.0f}-{max(ems):.0f}%")

File: analysis/revision/scripts/threshold_fig.R
# R port of threshold_fig.py (revision sensitivity analysis, cjfr-2026-0202).
# Run from the repo root: Rscript analysis/revision/scripts/threshold_fig.R
source("analysis/revision/scripts/revision_common.R")

m <- load_merged()
b <- build_pc1(m, "species")
pc <- b$pc1
pc_pooled <- build_pc1(m, "pooled")$pc1
m$struct <- m$percent_damaged > 1

mu <- mean(pc); med <- median(pc); sdv <- sd(pc)
t_otsu <- otsu_threshold(pc)
t_gmm <- gmm_1d(pc)$crossover
rules <- list("mean (published)" = list(t = mu,     col = "#c0392b"),
              "median"           = list(t = med,    col = "#2980b9"),
              "Otsu break"       = list(t = t_otsu, col = "#8e44ad"),
              "GMM crossover"    = list(t = t_gmm,  col = "#27ae60"))

# Panel B sweep: 5th-95th percentile (numpy default = R quantile type 7)
ts <- seq(quantile(pc, 0.05, type = 7), quantile(pc, 0.95, type = 7), length.out = 200)
mask_bgs <- m$site == "BGS"; mask_ems <- m$site == "EMS"; ns <- !m$struct
bgs <- numeric(length(ts)); ems <- numeric(length(ts))
for (i in seq_along(ts)) {
  inc <- ns & (pc > ts[i])
  bgs[i] <- sum(inc[mask_bgs]) / sum(mask_bgs) * 100
  ems[i] <- sum(inc[mask_ems]) / sum(mask_ems) * 100
}

# ---------- FIGURE ----------
png(file.path(OUT_DIR, "CJFR-threshold-fig.png"),
    width = 3282, height = 840, res = 175, pointsize = 10)
par(mfrow = c(1, 3), mar = c(4, 4, 3.5, 1), mgp = c(2.4, 0.7, 0))

# Panel A: density with candidate thresholds
xs2 <- seq(min(pc) - 0.4, max(pc) + 0.4, length.out = 300)
kd <- density(pc)
ky <- approx(kd$x, kd$y, xs2, rule = 2)$y
brk <- seq(min(pc), max(pc), length.out = 16)     # 15 bins over the data range
h <- hist(pc, breaks = brk, plot = FALSE)
plot(NA, xlim = range(xs2), ylim = c(0, 0.42), xaxs = "i", yaxs = "i",
     xlab = "ERT PC1  (higher = lower resistivity / more heterogeneous)",
     ylab = "density", main = "")
plot(h, freq = FALSE, col = "grey86", border = "white", add = TRUE)
lines(xs2, ky, col = "black", lwd = 2)
labs <- character(0); lcols <- character(0)
for (name in names(rules)) {
  r <- rules[[name]]
  if (is.finite(r$t)) {
    abline(v = r$t, col = r$col, lty = 2, lwd = 1.7)
    labs <- c(labs, sprintf("%s = %.2f", name, r$t))
    lcols <- c(lcols, r$col)
  }
}
legend("topright", legend = labs, col = lcols, lty = 2, lwd = 1.7, bty = "n", cex = 0.8)
EMDASH <- intToUtf8(0x2014)
title(main = "A", adj = 0, cex.main = 1, font.main = 2)

# Panel B: threshold sweep, BGS vs EMS incipient %
plot(NA, xlim = range(ts), ylim = range(c(bgs, ems)) + c(0, 4),
     xlab = "anomaly threshold on ERT PC1",
     ylab = "% of site trees classified \"incipient\"", main = "")
keep <- bgs >= ems
polygon(c(ts[keep], rev(ts[keep])), c(bgs[keep], rev(ems[keep])),
        col = adjustcolor("#1f77b4", alpha.f = 0.10), border = NA)
lines(ts, bgs, col = "#1f77b4", lwd = 2.4)
lines(ts, ems, col = "#d9822b", lwd = 2.4)
for (name in names(rules)) {
  r <- rules[[name]]
  if (is.finite(r$t)) abline(v = r$t, col = r$col, lty = 3, lwd = 1.3)
}
legend("topright", legend = c("BGS (wetland)", "EMS (upland)"),
       col = c("#1f77b4", "#d9822b"), lwd = 2.4, bty = "n", cex = 0.9)
title(main = "B", adj = 0, cex.main = 1, font.main = 2)

# Panel C: same distribution under the pooled normalization (preserves
# absolute between-site/species differences); replaces the former
# standalone normalization figure (CJFR-threshold-rules.png, archived).
kd3 <- density(pc_pooled)
xs3 <- seq(min(pc_pooled) - 0.4, max(pc_pooled) + 0.4, length.out = 300)
ky3 <- approx(kd3$x, kd3$y, xs3, rule = 2)$y
brk3 <- seq(min(pc_pooled), max(pc_pooled), length.out = 16)
h3 <- hist(pc_pooled, breaks = brk3, plot = FALSE)
plot(NA, xlim = range(xs3), ylim = c(0, max(h3$density, ky3) * 1.08),
     xaxs = "i", yaxs = "i",
     xlab = "ERT PC1, pooled normalization",
     ylab = "density", main = "")
plot(h3, freq = FALSE, col = "grey86", border = "white", add = TRUE)
lines(xs3, ky3, col = "black", lwd = 2)
abline(v = mean(pc_pooled), col = "#c0392b", lty = 2, lwd = 1.7)
legend("topright", legend = sprintf("mean = %.2f", mean(pc_pooled)),
       col = "#c0392b", lty = 2, lwd = 1.7, bty = "n", cex = 0.8)
title(main = "C", adj = 0, cex.main = 1, font.main = 2)
invisible(dev.off())

cat("saved. BGS never below EMS across sweep:",
    ifelse(all(bgs >= ems), "True", "False"), "\n")
cat(sprintf("BGS incipient range across sweep: %.0f-%.0f%% ; EMS: %.0f-%.0f%%\n",
            min(bgs), max(bgs), min(ems), max(ems)))
